File: translations/CJBench/J2P/MistralCommented/starcoder/Code/CRCAlgorithm.py

import random

class Main:
    # Declare instance variables
    def __init__(self, divisor, messageSize, ber):
        # Initialize instance variables
        self.correctMess = 0
        self.wrongMess = 0
        self.wrongMessCaught = 0
        self.wrongMessNotCaught = 0
        self.messSize = messageSize
        self.ber = ber
        self.messageChanged = False
        self.message = []
        self.p = []

        # Initialize p list with the given divisor string
        for i in range(len(divisor)):
            self.p.append(int(divisor[i]))

        # Initialize counter variables
        self.correctMess = 0
        self.wrongMess = 0
        self.wrongMessCaught = 0
        self.wrongMessNotCaught = 0
        self.ber = ber

    # Getters for counter variables
    def getWrongMess(self):
        return self.wrongMess

    def getWrongMessCaught(self):
        return self.wrongMessCaught

    def getWrongMessNotCaught(self):
        return self.wrongMessNotCaught

    def getCorrectMess(self):
        return self.correctMess

    # Refactor method to reset instance variables
    def refactor(self):
        self.messageChanged = False
        self.message = []

    # Generate random message method
    def generateRandomMess(self):
        # Generate random message of given size
        for i in range(self.messSize):
            x = random.randint(0, 1)
            self.message.append(x)

    # Divide message with p method without checking
    def divideMessageWithP(self, check):
        x = [] # Initialize x list
        k = self.message.copy() # Initialize k list as a copy of message list

        # If not checking, fill k list with zeros
        if not check:
            for i in range(len(self.p) - 1):
                k.append(0)

        # Divide message with p
        while len(k) > 0:
            while len(x) < len(self.p) and len(k) > 0:
                x.append(k.pop(0))

            # If message size is equal to p size
            if len(x) == len(self.p):
                # Xor each element in x list with corresponding element in p list
                for i in range(len(self.p)):
                    x[i] = x[i] ^ self.p[i]

                # Remove leading zeros from x list
                while len(x) > 0 and x[0] == 0:
                    x.pop(0)
            else:
                break

        # Update counter variables based on the result
        if not check:
            self.message.extend(x)
        else:
            # If message has been changed and 1 is present in x list
            if self.messageChanged and 1 in x:
                self.wrongMessCaught += 1
            # If message has been changed and 1 is not present in x list
            elif self.messageChanged and 1 not in x:
                self.wrongMessNotCaught += 1
            # If message has not been changed
            else:
                self.correctMess += 1

    # Change message method
    def changeMess(self):
        # Change each element in message list with probability ber
        for i in range(len(self.message)):
            if random.random() < self.ber:
                self.messageChanged = True
                self.message[i] = self.message[i] ^ 1

        # Update counter variable if message has been changed
        if self.messageChanged:
            self.wrongMess += 1

# Main method to run the simulation
if __name__ == "__main__":
    # Read input from user
    divisor = input()
    messageSize = int(input())
    ber = float(input())
    numMessages = int(input())

    # Create Main object with given input
    crc = Main(divisor, messageSize, ber)

    # Generate and check messages for given number of times
    for i in range(numMessages):
        crc.refactor() # Refactor before each iteration
        crc.generateRandomMess() # Generate random message
        crc.divideMessageWithP(False) # Divide message with p without checking
        crc.changeMess() # Change message
        crc.divideMessageWithP(True) # Divide message with p and check

    # Print the results
    print("Correct Messages: " + str(crc.getCorrectMess()))
    print("Wrong Messages: " + str(crc.getWrongMess()))
    print("Wrong Messages Caught: " + str(crc.getWrongMessCaught()))
    print("Wrong Messages Not Caught: " + str(crc.getWrongMessNotCaught()))

File: datasets/CJBench/Java/MistralCommented/CRCAlgorithm.java

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.concurrent.ThreadLocalRandom;

public class Main {
    // Declare instance variables
    private int correctMess;
    private int wrongMess;
    private int wrongMessCaught;
    private int wrongMessNotCaught;
    private int messSize;
    private double ber;
    private boolean messageChanged;
    private List<Integer> message;
    private List<Integer> p;

    public Main(String divisor, int messageSize, double ber) {
        // Initialize instance variables
        this.messageChanged = false;
        this.message = new ArrayList<>();
        this.messSize = messageSize;
        this.p = new ArrayList<>();

        // Initialize p list with the given divisor string
        for (int i = 0; i < divisor.length(); i++) {
            p.add(Character.getNumericValue(divisor.charAt(i)));
        }

        // Initialize counter variables
        this.correctMess = 0;
        this.wrongMess = 0;
        this.wrongMessCaught = 0;
        this.wrongMessNotCaught = 0;
        this.ber = ber;
    }

    // Getters for counter variables
    public int getWrongMess() {
        return wrongMess;
    }

    public int getWrongMessCaught() {
        return wrongMessCaught;
    }

    public int getWrongMessNotCaught() {
        return wrongMessNotCaught;
    }

    public int getCorrectMess() {
        return correctMess;
    }

    // Refactor method to reset instance variables
    public void refactor() {
        this.messageChanged = false;
        this.message = new ArrayList<>();
    }

    // Generate random message method
    public void generateRandomMess() {
        // Generate random message of given size
        for (int i = 0; i < messSize; i++) {
            int x = ThreadLocalRandom.current().nextInt(0, 2);
            message.add(x);
        }
    }

    // Divide message with p method without checking
    public void divideMessageWithP(boolean check) {
        List<Integer> x = new ArrayList<>(); // Initialize x list
        List<Integer> k = new ArrayList<>(message); // Initialize k list as a copy of message list

        // If not checking, fill k list with zeros
        if (!check) {
            for (int i = 0; i < p.size() - 1; i++) {
                k.add(0);
            }
        }

        // Divide message with p
        while (!k.isEmpty()) {
            while (x.size() < p.size() && !k.isEmpty()) {
                x.add(k.remove(0));
            }

            // If message size is equal to p size
            if (x.size() == p.size()) {
                // Xor each element in x list with corresponding element in p list
                for (int i = 0; i < p.size(); i++) {
                    x.set(i, x.get(i) ^ p.get(i));
                }

                // Remove leading zeros from x list
                while (!x.isEmpty() && x.get(0) == 0) {
                    x.remove(0);
                }
            }
        }

        // Update counter variables based on the result
        if (!check) {
            message.addAll(x);
        } else {
            // If message has been changed and 1 is present in x list
            if (messageChanged && x.contains(1)) {
                wrongMessCaught++;
            }
            // If message has been changed and 1 is not present in x list
            else if (messageChanged && !x.contains(1)) {
                wrongMessNotCaught++;
            }
            // If message has not been changed
            else if (!messageChanged) {
                correctMess++;
            }
        }
    }

    // Change message method
    public void changeMess() {
        // Change each element in message list with probability ber
        for (int i = 0; i < message.size(); i++) {
            if (Math.random() < ber) {
                messageChanged = true;
                message.set(i, message.get(i) ^ 1);
            }
        }

        // Update counter variable if message has been changed
        if (messageChanged) {
            wrongMess++;
        }
    }

    // Main method to run the simulation
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Read input from user
        String divisor = scanner.next();
        int messageSize = scanner.nextInt();
        double ber = scanner.nextDouble();
        int numMessages = scanner.nextInt();

        // Create Main object with given input
        Main crc = new Main(divisor, messageSize, ber);

        // Generate and check messages for given number of times
        for (int i = 0; i < numMessages; i++) {
            crc.refactor(); // Refactor before each iteration
            crc.generateRandomMess(); // Generate random message
            crc.divideMessageWithP(false); // Divide message with p without checking
            crc.changeMess(); // Change message
            crc.divideMessageWithP(true); // Divide message with p and check
        }

        // Print the results
        System.out.println("Correct Messages: " + crc.getCorrectMess());
        System.out.println("Wrong Messages: " + crc.getWrongMess());
        System.out.println("Wrong Messages Caught: " + crc.getWrongMessCaught());
        System.out.println("Wrong Messages Not Caught: " + crc.getWrongMessNotCaught());

        scanner.close();
    }
}